xlml/apis/task.py

```diff
@@ -503,3 +504,118 @@ def clean_up(self, resource: airflow.XComArg, project_id: str, zone: str) -> DAG
       AirflowTaskTimeout: An error occurs when execution_timeout is breached.
     """
     return gpu.delete_resource.override(group_id="clean_up")(resource, project_id, zone)
+
+
+@dataclasses.dataclass
+class GpuGkeTask(BaseTask):
+  """This is a class to set up tasks for GPU.
+
+  Attributes:
+    image_project: the project that an image belongs to.
+    image_family: the family group that an image belongs to.
+  """
+
+  task_test_config: test_config.JSonnetGpuTest
+  task_gcp_config: gcp_config.GCPConfig
+  cluster_name: str
+  job_create_timeout: datetime.timedelta = datetime.timedelta(minutes=10)
+  # TODO: metrics
+  # task_metric_config: Optional[metric_config.MetricConfig] = None
+
+  def run(self) -> DAGNode:
+    """Run a test job.
+
+    Returns:
+      A task group with the following tasks chained: provision, run_model,
+      post_process, clean_up.
+    """
+    # piz: We skip the queued resource for GPU for now since there is no queued
+    # resource command for GPU.
+    with TaskGroup(
+        group_id=self.task_test_config.benchmark_id, prefix_group_id=True
+    ) as group:
+      job_body = self._get_job_manifest()
+      gke.run_job(
+          job_body, self.task_gcp_config, self.cluster_name, self.job_create_timeout
+      )
+
+    return group
+
+  def _get_job_manifest(self):
+    return {
```

Comment:
Can we move those to a config/manifest file and read/import here?

Reply:
Ironically, since this is templated based parameters of the job config, the best place to move this as  static config would be JSonnet :laughing: 

I'd like to keep as much "logic" in Python compared to JSonnet moving forward, and this makes the most sense as a private helper since it depends only on the attributes of this class. But `task.py` is getting untenably long and hard to read, and this is a particularly long method because Kubernetes manifests are verbose. I propose two options:

1. Move each task type (or group of similar task types) to new files like `task.gke` or `task.gpu`.
2. Move each task type to the place where their implementations are defined (e.g. this becomes `gke.GpuGkeTask`.

Reply:
Yes, this is really long now. I may prefer the 1st option, so we will have:

based on accelerators:
```
task_gpu.py
task_tpu.py
``` 

Or based on platforms:

```
task_gke.py
task_gce.py
```

This will be a big refactor, can you add a TODO with my name.